### PR TITLE
GEOMESA-2614 RoutedDataStoreView impl

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Combined data store view that routes queries to one of the underling stores

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by elahrvivaz** [comment]
stacks on top of #2311

File: docs/user/ds_views.rst [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        }
+      ]
+    }
+
+Custom Routing
+""""""""""""""
+
+As an alternative to routing by attribute, routing can be delegated to a provider interface:
+``org.locationtech.geomesa.index.view.RouteSelector``. Implementations of this class must be made available via
+Java `SPI loading <http://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html>`__, with a special
```
Comment (jnh5y) [inline docs/user/ds_views.rst]: We should reference the Java 8 docs for SPI;).

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelector.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  def init(stores: Seq[(DataStore, java.util.Map[String, AnyRef])]): Unit
+
+  /**
+    * Route a query to a particular store. If no store is selected, query will return empty
+    *
+    * @param sft simple feature type
+    * @param query query
+    * @return
+    */
+  def route(sft: SimpleFeatureType, query: Query): Option[DataStore]
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelector.scala]: Do we care about scala.Option being in the API?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelectorByAttribute.scala [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+  override def init(stores: Seq[(DataStore, java.util.Map[String, AnyRef])]): Unit = {
+    val builder = Seq.newBuilder[(Set[String], DataStore)]
+    stores.foreach { case (ds, config) =>
+      Option(config.get(RouteAttributes)).foreach { attributes =>
+        try {
+          attributes.asInstanceOf[java.util.List[AnyRef]].asScala.foreach {
+            case s: String if s.equalsIgnoreCase("id") => id = Some(ds)
+            case s: String => builder += Set(s) -> ds
+            case list: java.util.List[String] if list.isEmpty => include = Some(ds)
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelectorByAttribute.scala]: If I'm reading this right, in the case of multiple stores being configured for Filter.INCLUDE, only the last is used?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreView.scala [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+/**
+  * Routed querying against multiple data stores
+  *
+  * @param stores delegate stores
+  * @param router query router
+  * @param namespace schema namespace
+  */
+class RoutedDataStoreView(val stores: Seq[DataStore], val router: RouteSelector, namespace: Option[String])
+    extends MergedDataStoreSchemas(stores, namespace) with HasGeoMesaStats {
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreView.scala]: Ok, what's the schema namespace?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreViewFactory.scala [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      }
+    } catch {
+      case NonFatal(e) => stores.result.foreach(_._1.dispose()); throw e
+    }
+    val storesAndConfigs = stores.result
+
+    val router = {
+      val impl = RouterParam.lookup(params)
+      ServiceLoader.load[RouteSelector]().find(_.getClass.getName == impl).getOrElse {
+        throw new IllegalArgumentException(s"Could not load route selector '$impl'. Available implementations: " +
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreViewFactory.scala]: If we die here, should we clean up the DataStores which have been instantiated?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedFeatureSourceView.scala [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  override def getDataStore: DataStore = ds
+
+  override def getSupportedHints: java.util.Set[Key] = hints
+
+  override def getInfo: ResourceInfo = new DelegatingResourceInfo(this)
+
+  override def addFeatureListener(listener: FeatureListener): Unit = throw new NotImplementedError()
+
+  override def removeFeatureListener(listener: FeatureListener): Unit = throw new NotImplementedError()
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedFeatureSourceView.scala]: Is this consistent with what we do elsewhere?  I'm wondering if a No-op would be better.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/package.scala [hunk 7]
```diff
@@ -45,4 +52,100 @@ package object view {
...
+    override def getFeatureWriterAppend(typeName: String, transaction: Transaction): SimpleFeatureWriter = error
+  }
+
+  /**
+    * Data store that merges the schemas from multiple delegate stores and presents them as a unified result
+    *
+    * @param stores delegate stores
+    * @param namespace schema namespace
+    */
+  abstract class MergedDataStoreSchemas(stores: Seq[DataStore], namespace: Option[String])
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/package.scala]: Are there other MDSS which we could be implementing?

File: docs/user/ds_views.rst (hunk 1 above)
Comment (elahrvivaz) [reply docs/user/ds_views.rst]: fixed

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelector.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelector.scala]: I mean, I don't care... do you? ;)

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelectorByAttribute.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelectorByAttribute.scala]: yeah, that would be an error case - I can add a check for that

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreView.scala (hunk 4 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreView.scala]: that's just the namespace configured via geoserver, used to namespace simple feature types in getSchema

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreViewFactory.scala (hunk 5 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedDataStoreViewFactory.scala]: good catch, fixed

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedFeatureSourceView.scala (hunk 6 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RoutedFeatureSourceView.scala]: This is consistent with what we do elsewhere, e.g. https://github.com/locationtech/geomesa/blob/master/geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureSource.scala#L101-L103. Not saying it's correct, per say...

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/package.scala (hunk 7 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/package.scala]: So far just the two :)

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelectorByAttribute.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelectorByAttribute.scala]: fixed

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelector.scala (hunk 2 above)
Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelector.scala]: Do we expect any Java implementations?

Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/view/RouteSelector.scala]: so far my experience has been we make java apis and then no one ever uses them except us, in scala :)

**Pull request merged**